pulser-core/pulser/noise_model.py

```diff
@@ -113,115 +158,203 @@ class NoiseModel:
+            laser_waist=laser_waist,
+            amp_sigma=amp_sigma,
+            relaxation_rate=relaxation_rate,
+            dephasing_rate=dephasing_rate,
+            hyperfine_dephasing_rate=hyperfine_dephasing_rate,
+            depolarizing_rate=depolarizing_rate,
+            eff_noise_rates=to_tuple(eff_noise_rates),
+            eff_noise_opers=to_tuple(eff_noise_opers),
+        )
+
+        if noise_types is not None:
+            with warnings.catch_warnings():
+                warnings.simplefilter("always")
+                warnings.warn(
+                    "The explicit definition of noise types is deprecated; "
+                    "doing so will use legacy default values for all relevant "
+                    "parameters that are not given a custom value. Instead, "
+                    "defining only the necessary parameters is reccomended; "
+                    "doing so (when the noise types are not explicitly given) "
+                    "will disregard all undefined parameters.",
+                    DeprecationWarning,
+                    stacklevel=2,
+                )
+            self._check_noise_types(noise_types)
+            for nt_ in noise_types:
+                for p_ in _NOISE_TYPE_PARAMS[nt_]:
+                    # Replace undefined relevant params by the legacy default
+                    if param_vals[p_] is None:
+                        param_vals[p_] = _LEGACY_DEFAULTS[p_]
+
+        true_noise_types: set[NoiseTypes] = {
+            cast(NoiseTypes, _PARAM_TO_NOISE_TYPE[p_])
+            for p_ in param_vals
+            if param_vals[p_] and p_ in _PARAM_TO_NOISE_TYPE
+        }
+
+        self._check_eff_noise(
+            cast(tuple, param_vals["eff_noise_rates"]),
+            cast(tuple, param_vals["eff_noise_opers"]),
+            "eff_noise" in (noise_types or true_noise_types),
+        )
+
+        # Get rid of unnecessary None's
+        for p_ in _POSITIVE | _PROBABILITY_LIKE:
+            param_vals[p_] = param_vals[p_] or 0.0
+
+        relevant_params = self._find_relevant_params(
+            true_noise_types,
+            cast(float, param_vals["state_prep_error"]),
+            cast(float, param_vals["amp_sigma"]),
+            cast(Union[float, None], param_vals["laser_waist"]),
+        )
+
+        if noise_types is not None:
+            if true_noise_types != set(noise_types):
+                raise ValueError(
+                    "The explicit definition of noise types (deprecated) is"
+                    " not compatible with the modification of unrelated noise "
+                    "parameters. Defining only the relevant noise parameters "
+                    "(without specifying the noise types) is recommended."
                 )
+            run_params_ = [
+                p for p in relevant_params if p in ("runs", "samples_per_run")
+            ]
+            for p_ in run_params_:
+                param_vals[p_] = param_vals[p_] or _LEGACY_DEFAULTS[p_]
```

Comment:
I must say I got lost and can't really understand the need of this for loop - why do we end up using LEGACY_DEFAULTS ? I thought we got rid of None...

Reply:
- We got rid of `None` in quantities that are not strictly positive (which is not the case for the run parameters)
- Here we rely on the legacy default for the run parameters in case `noise_types` was explicitly given **and** the run parameters are relevant. It is done here because we can only know if the run parameters are relevant after we know the values of `state_prep_error` and `amp_sigma` (which we potentially filled with the legacy defaults above)

I'll add a comment